### ISSUE TITLE
PathGenerator - Optimization Attempt.

### ISSUE THREAD

**Issue by Langerz82** [TrinityCore/TrinityCore#22116]
**Description:**

Attempt at optimizing the PathGenerator. I have no idea if this is gonna work and/or it really saves on processing.
See File:
https://gist.github.com/Langerz82/0e043c122e11c401e316351faa7877dd

**Current behaviour:**

PathGenerator consumes allot of processing.

**Expected behaviour:**

PathGenerator should be running optimally.

**Steps to reproduce the problem:**

No Steps.

**Branch(es):**

3.3.5

**TC rev. hash/commit:** 

TrinityCore rev. 2e6100919d77+ 2018-06-25 00:42:30 +0200 (3.3.5 branch) (Win64, Debug, Static) (worldserver-daemon)

**Operating system:** Win 10

**Comment by Langerz82** [comment]
I made some minor changes, please re-test. I tested briefly in dungeons and with obstacles and it seems to be doing a pretty good job at making the right path.

**Comment by Shauren** [comment]
> I have no idea if this is gonna work and/or it really saves on processing.

This invalidates all your "attempts" - use proper profiling tools instead of optimizing on "feelings"

**Comment by Langerz82** [comment]
I was hoping someone more experienced could do that, as I'm noob at profiling.

**Comment by Langerz82** [comment]
Updated code, as some Mobs were jamming in stair hallways and added variables to indicate that the players returned path is not the complete one but only a partial.

See changes:
https://gist.github.com/Langerz82/0e043c122e11c401e316351faa7877dd

**Comment by jackpoz** [comment]
Your description doesn't really explain much about what you changed in your code and which cases you are trying to optimize, could you be more specific ? We usually have PRs when dealing with C++ changes, it helps the code review

**Comment by Langerz82** [comment]
ok, let me briefly explain the code.

It first does a raycast to the end point and detects if there is a collision. If it's close to a mmap border, or too far away to the border then it will build the path as normal.

If it meets the condition it does a second check to return the point of collision and builds a sub path to the collision and returns that instead. I haven't checked it for complex movements, but I tested it on a angular spiral staircase with mobs and they chase and return to there original positions as desired. It needs thorough testing but from my preliminary tests it looked promising.

My justification for using sub paths is we do not need to calculate the full path every path cycle.

I tried profiling but the function never came up in CodeXL so im not sure exactly how to measure the optimization boost properly and hoping that you guys could do so.

**Comment by AlexBolotsin** [comment]
@jackpoz Does TC have some profiling hooks or everything? Maybe there is a system that could return spent time statistics? Should we use some 3rd party solutions for that ie gprof?

**Comment by Shauren** [comment]
There are no profiling utilities in the code, you should use tools like gprof or visual studio profiler

**Comment by Langerz82** [comment]
Rochet2 provided a good testing time example in the other case so I will make some operational statistics with it.

**Comment by Langerz82** [comment]
It's so strange, I'll test both in runtime of the boot process and report the results.

**Comment by Langerz82** [comment]
I'm debugging and testing times. I've noticed a few issues when the sub end point has no collision for example. It's a work in progress.

**Comment by Langerz82** [comment]
Yep currently the patch makes no real difference. I'm fiddling with the pathfinding to see if I can improve on it while maintaining accuracy. The big cost I've seen so far is finding the z-axis of the map.

**Comment by Langerz82** [comment]
Here is what I got so far for anybody interested.

See:
https://gist.github.com/Langerz82/0e043c122e11c401e316351faa7877dd

**Comment by Langerz82** [comment]
Does anyone know by looking at this code why it throws the dtResult flag DT_BUFFER_TOO_SMALL, t.maxpath is a large negative value.

```diff
void PathGenerator::NormalizePath()
{
    // No need to Normalize first path as it is expensive to do so.
    for (uint32 i = 0; i < _pathPoints.size(); ++i)
    {
        if (_sourceUnit)
        {
            if (!_navMeshQuery && !_navMesh)
                _sourceUnit->UpdateAllowedPositionZ(_pathPoints[i].x, _pathPoints[i].y, _pathPoints[i].z);
            else
            {
                const float point[VERTEX_SIZE] = { _pathPoints[i].y, _pathPoints[i].z, _pathPoints[i].x};
                const float point2[VERTEX_SIZE] = { _pathPoints[i].y, _pathPoints[i].z-50.f, _pathPoints[i].x};
                float dist = 0;
                dtPolyRef polyRef = GetPolyByLocation(point, &dist);
                if (polyRef)
                {
                    dtRaycastHit hit;
                    dtStatus dtResult = _navMeshQuery->raycast(
                        polyRef,
                        point,
                        point2,
                        &_filter,
                        0,
                        &hit,
                        polyRef);
                    if ((dtResult & DT_SUCCESS) && hit.t > 0.0f && hit.t < 50.f)
                    {
                        _pathPoints[i].z = _pathPoints[i].z + hit.t;
                    }
                }
            }
        }
    }
}
```

**Comment by Langerz82** [comment]
Changing the Normalize Path to use mmaps made it sample at 300microseconds roughly. :x

**Comment by Langerz82** [comment]
See:
https://gist.github.com/Langerz82/13e76058a5dada990d8c06d795b22c93

**Comment by Langerz82** [comment]
Sorry, shouldnt have closed this as that patch I submitted above want committed.

**Comment by illfated** [comment]
Pull Request is usually required for C++ code changes to be committed.

**Comment by Langerz82** [comment]
I'll do a PR when it tested a bit more, I also wana make the chase recalc a configurable value in worldserver.conf which I still need to do.
